src/net.cpp

```diff
@@ -2985,3 +2985,27 @@ uint64_t CConnman::CalculateKeyedNetGroup(const CAddress& ad) const
 
     return GetDeterministicRandomizer(RANDOMIZER_ID_NETGROUP).Write(vchNetGroup.data(), vchNetGroup.size()).Finalize();
 }
+
+void CaptureMessage(const CAddress& addr, const std::string& msg_type, const Span<const unsigned char>& data, bool is_incoming)
+{
+    auto time = GetTime<std::chrono::microseconds>();
+
+    // Windows folder names can not include a colon
+    std::string clean_addr = addr.ToString();
+    std::replace(clean_addr.begin(), clean_addr.end(), ':', '_');
+
+    fs::path base_path = GetDataDir() / "message_capture" / clean_addr;
+    fs::create_directories(base_path);
+
+    fs::path path = base_path / (is_incoming ? "msgs_recv.dat" : "msgs_sent.dat");
+    CAutoFile f(fsbridge::fopen(path, "ab"), SER_DISK, CLIENT_VERSION);
```

Comment:
2dd63707380ae5cdf79a2a8d98e0133f881f9381

I know this is a debugging feature, but file open/close for each message is kind of bad right, also because close flushes?

Reply:
I'm not sure that I see the disadvantage.  The alternative required using a greater number of file descriptors, which was deemed to be worse.

Reply:
If measurements show that this ever becomes a problem in practice it can be tackled in a follow-up.

TBH I'm much more worried about the possible file descriptor DoS vector that we would risk open up if the file descriptor were left open open.

Security first! :)